src/coreComponents/mesh/generators/VTKMeshGenerator.cpp

```diff
@@ -0,0 +1,693 @@
+}
+
+void VTKMeshGenerator::generateMesh( DomainPartition & domain )
+{
+  vtkSmartPointer<vtkMultiProcessController> controller = GetVTKController();
+  vtkMultiProcessController::SetGlobalController(controller);
+
+  vtkSmartPointer<vtkUnstructuredGrid> loadedMesh = LoadVTKMesh();
+  MpiWrapper::barrier();
+
+  RedistributeMesh(loadedMesh);
+  MpiWrapper::barrier();
+
+  GEOSX_LOG_RANK_0(" Successfully redistributed the mesh on the " << controller->GetNumberOfProcesses() << " MPI processes");
+
+  Group & meshBodies = domain.getGroup( string( "MeshBodies" ));
+  MeshBody & meshBody = meshBodies.registerGroup< MeshBody >( this->getName() );
+
+  MeshLevel & meshLevel0 = meshBody.registerGroup< MeshLevel >( string( "Level0" ));
+  NodeManager & nodeManager = meshLevel0.getNodeManager();
+
+  double globalLength = WriteMeshNodes( nodeManager );
+  meshBody.setGlobalLengthScale( globalLength);
+
+  ComputePotentialNeighborLists( domain );
+
+  localIndex nbHex =0;
+  localIndex nbTet = 0;
+  localIndex nbWedge = 0;
+  localIndex nbPyr = 0;
+
+  // region id link to their number of cells
+  std::map<int,localIndex> regions_hex;
+  std::map<int,localIndex> regions_tetra;
+  std::map<int,localIndex> regions_wedges;
+  std::map<int,localIndex> regions_pyramids;
+
+  // surface and greionids
+  std::set<int> regions;
+  std::set<int> surfaces; //local to this rank
+  std::vector<int> allSurfaces; // global to the whole mesh
+
+  CountCellsAndFaces( nbHex, nbTet, nbWedge, nbPyr, regions_hex, regions_tetra, regions_wedges, regions_pyramids, regions, surfaces, allSurfaces);
+
+  std::vector< vtkDataArray * > arraysToBeImported = FindArrayToBeImported();
+
+  WriteCellBlocks( domain, nbHex, nbTet, nbWedge, nbPyr,regions_hex, regions_tetra, regions_wedges, regions_pyramids, regions, arraysToBeImported );
+
+  WriteSurfaces( nodeManager, allSurfaces);
+
+  GEOSX_LOG_RANK_0("Mesh was loaded successfully");
+}
+
+vtkSmartPointer<vtkMultiProcessController> VTKMeshGenerator::GetVTKController()
+{
+  #ifdef GEOSX_USE_MPI
+    vtkNew<vtkMPIController> controller;
+    vtkMPICommunicatorOpaqueComm vtkGeosxComm(&MPI_COMM_GEOSX);
+    vtkNew<vtkMPICommunicator> communicator;
+    communicator->InitializeExternal( &vtkGeosxComm);
+    controller->SetCommunicator(communicator);
+  #else
+    vtkNew<vtkDummyController> controller;
+  #endif
+  return controller;
+}
+
+vtkSmartPointer<vtkUnstructuredGrid> VTKMeshGenerator::LoadVTKMesh()
+{
+  string_array filePathTokenized = stringutilities::tokenize( m_filePath, "."); //TODO maybe code a method in Path to get the file extension?
+  string extension = filePathTokenized[filePathTokenized.size() - 1];
+  vtkSmartPointer< vtkUnstructuredGrid > loadedMesh = vtkSmartPointer< vtkUnstructuredGrid >::New();
+  if( extension == "pvtu" )
+  {
+    vtkSmartPointer<vtkXMLPUnstructuredGridReader> vtkUgReader = vtkSmartPointer<vtkXMLPUnstructuredGridReader>::New();
```

Comment:
I prefer letting it like this because it allows to control the file format I can import. A generic reader exists but it words for the whole VTK file formats...

Not sure for the lambda, I find them counter intuitive to read

Reply:
The problem is the duplication of the 
```c++
vtkUgReader->SetFileName( m_filePath.c_str() );
vtkUgReader->Update();
loadedMesh = vtkUgReader->GetOutput();
```
lines.
If there is a base class that helps generic coding (being in fact pointers to derived class; polymorphism), it should be OK and you can control the file format nevertheless.

Reply:
Yes I see. I proposed something with a lambda (the generic reader is different for the XML files and the legacy ones so...)